yb-voyager/src/queryparser/query_parser.go

```diff
@@ -0,0 +1,90 @@
+package queryparser
+
+import (
+	"fmt"
+
+	pg_query "github.com/pganalyze/pg_query_go/v5"
+	log "github.com/sirupsen/logrus"
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+const (
+	PG_QUERY_NODE_NODE      = "pg_query.Node"
+	PG_QUERY_STRING_NODE    = "pg_query.String"
+	PG_QUERY_ASTAR_NODE     = "pg_query.A_Star"
+	PG_QUERY_XMLEXPR_NODE   = "pg_query.XmlExpr"
+	PG_QUERY_FUNCCALL_NODE  = "pg_query.FuncCall"
+	PG_QUERY_COLUMNREF_NODE = "pg_query.ColumnRef"
+)
+
+type QueryParser struct {
+	QueryString string
+	ParseTree   *pg_query.ParseResult
+}
+
+func New(query string) *QueryParser {
+	return &QueryParser{
+		QueryString: query,
+	}
+}
+
+func (qp *QueryParser) Parse() error {
+	log.Debugf("parsing the query-%s", qp.QueryString)
+	tree, err := pg_query.Parse(qp.QueryString)
+	if err != nil {
+		return err
+	}
+	qp.ParseTree = tree
+	return nil
+}
+
+func (qp *QueryParser) GetUnsupportedQueryConstructs() ([]utils.UnsupportedQueryConstruct, error) {
+	var result []utils.UnsupportedQueryConstruct = nil
+	visited := make(map[protoreflect.Message]bool)
+	unsupportedConstructs := make(map[string]bool) // set for storing unique constructs
+
+	log.Debugf("Query: %s\n", qp.QueryString)
+	log.Debugf("ParseTree: %+v\n", qp.ParseTree)
+	detectors := []UnsupportedConstructDetector{
+		NewFuncCallDetector(),
+		NewColumnRefDetector(),
+		NewXmlExprDetector(),
+	}
+
+	compositeDetector := &CompositeDetector{detectors: detectors}
```

Comment:
As discussed, this seems fancy and nice, but IMO, unnecessary complication 😄 
Would suggest just looping over the detectors in your processor function

Reply:
ok will remove it. It can be complicated to understand..